From b3a4eba6854a2fa53a49d2ecc11c45e6c9745ac7 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:03 -0400
Subject: [fs] btrfs: fix possible memory leaks in open_ctree()

Message-id: <1407179109-13714-297-git-send-email-sandeen@redhat.com>
Patchwork-id: 87315
O-Subject: [RHEL7.1 PATCH 296/422] Btrfs: fix possible memory leaks in open_ctree()
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Wang Shilong <wangsl.fnst@cn.fujitsu.com>

commit 28c16cbbc32781224309e50cc99c684f2498bc59
Author: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
Date:   Wed Apr 23 19:33:35 2014 +0800

    Btrfs: fix possible memory leaks in open_ctree()

    Fix possible memory leaks in the following error handling paths:

    read_tree_block()
    btrfs_recover_log_trees
    btrfs_commit_super()
    btrfs_find_orphan_roots()
    btrfs_cleanup_fs_roots()

    Signed-off-by: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index 54e2434..5c52edb 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -2863,7 +2863,7 @@ retry_root_backup:
    printk(KERN_ERR "BTRFS: failed to read log tree\n");
    free_extent_buffer(log_tree_root->node);
    kfree(log_tree_root);
-   goto fail_trans_kthread;
+   goto fail_qgroup;
   }
   /* returns with log_tree_root freed on success */
   ret = btrfs_recover_log_trees(log_tree_root);
@@ -2872,24 +2872,24 @@ retry_root_backup:
         "Failed to recover log tree");
    free_extent_buffer(log_tree_root->node);
    kfree(log_tree_root);
-   goto fail_trans_kthread;
+   goto fail_qgroup;
   }
 
   if (sb->s_flags & MS_RDONLY) {
    ret = btrfs_commit_super(tree_root);
    if (ret)
-    goto fail_trans_kthread;
+    goto fail_qgroup;
   }
  }
 
  ret = btrfs_find_orphan_roots(tree_root);
  if (ret)
-  goto fail_trans_kthread;
+  goto fail_qgroup;
 
  if (!(sb->s_flags & MS_RDONLY)) {
   ret = btrfs_cleanup_fs_roots(fs_info);
   if (ret)
-   goto fail_trans_kthread;
+   goto fail_qgroup;
 
   ret = btrfs_recover_relocation(tree_root);
   if (ret < 0) {
-- 
1.7.1